File: backend/src/db/mappers.ts
import type { Product, Purchase, ShardSchema, StoredProduct, StoredUser, Transaction, User } from "../types/market.js";

function toIso(value: string | Date) {
  return new Date(value).toISOString();
}

export function toNumber(value: string | number) {
  return typeof value === "number" ? value : Number(value);
}

export function mapUser(row: Record<string, unknown>, schema?: ShardSchema): User {
  return {
    id: String(row.id),
    fullName: String(row.full_name),
    email: String(row.email),
    balance: toNumber(row.balance as string | number),
    createdAt: toIso(row.created_at as string | Date),
    twoFactorEnabled: Boolean(row.two_factor_enabled)
  };
}

export function mapStoredUser(row: Record<string, unknown>, schema: ShardSchema): StoredUser {
  return {
    ...mapUser(row, schema),
    passwordHash: String(row.password_hash),
    schema
  };
}

export function mapProduct(row: Record<string, unknown>): Product {
  return {
    id: String(row.id),
    ownerId: String(row.owner_id),
    name: String(row.name),
    brand: String(row.brand),
    category: String(row.category),
    description: String(row.description),
    price: toNumber(row.price as string | number),
    quantity: Number(row.quantity),
    condition: row.condition as Product["condition"],
    status: row.status as Product["status"],
    listedAt: toIso(row.listed_at as string | Date),
    soldCount: Number(row.sold_count),
    color: String(row.color)
  };
}

export function mapStoredProduct(row: Record<string, unknown>, schema: ShardSchema): StoredProduct {
  return {
    ...mapProduct(row),
    schema
  };
}

export function mapPurchase(row: Record<string, unknown>): Purchase {
  return {
    id: String(row.id),
    productId: String(row.product_id),
    productName: String(row.product_name),
    buyerId: String(row.buyer_id),
    sellerId: String(row.seller_id),
    amount: toNumber(row.amount as string | number),
    purchasedAt: toIso(row.purchased_at as string | Date),
    status: row.status as Purchase["status"]
  };
}

export function mapTransaction(row: Record<string, unknown>): Transaction {
  return {
    id: String(row.id),
    type: row.type as Transaction["type"],
    amount: toNumber(row.amount as string | number),
    fromUserId: row.from_user_id ? String(row.from_user_id) : undefined,
    toUserId: row.to_user_id ? String(row.to_user_id) : undefined,
    productId: row.product_id ? String(row.product_id) : undefined,
    description: String(row.description),
    createdAt: toIso(row.created_at as string | Date),
    status: row.status as Transaction["status"]
  };
}

File: backend/src/main/java/com/marketplace/db/Mappers.java
package com.marketplace.db;

import com.marketplace.model.*;
import java.sql.ResultSet;
import java.sql.SQLException;

public class Mappers {

    public static String toIso(Object value) {
        if (value == null) return null;
        if (value instanceof java.sql.Timestamp) {
            return ((java.sql.Timestamp) value).toInstant().toString();
        }
        if (value instanceof java.time.OffsetDateTime) {
            return ((java.time.OffsetDateTime) value).toInstant().toString();
        }
        if (value instanceof java.util.Date) {
            return ((java.util.Date) value).toInstant().toString();
        }
        return value.toString();
    }

    public static double toDouble(Object value) {
        if (value == null) return 0.0;
        if (value instanceof Number) {
            return ((Number) value).doubleValue();
        }
        try {
            return Double.parseDouble(value.toString());
        } catch (NumberFormatException e) {
            return 0.0;
        }
    }

    public static User mapUser(ResultSet rs) throws SQLException {
        return new User(
            rs.getString("id"),
            rs.getString("full_name"),
            rs.getString("email"),
            toDouble(rs.getObject("balance")),
            toIso(rs.getObject("created_at")),
            rs.getBoolean("two_factor_enabled")
        );
    }

    public static StoredUser mapStoredUser(ResultSet rs, String schema) throws SQLException {
        User user = mapUser(rs);
        return new StoredUser(user, rs.getString("password_hash"), schema);
    }

    public static Product mapProduct(ResultSet rs) throws SQLException {
        return new Product(
            rs.getString("id"),
            rs.getString("owner_id"),
            rs.getString("name"),
            rs.getString("brand"),
            rs.getString("category"),
            rs.getString("description"),
            toDouble(rs.getObject("price")),
            rs.getInt("quantity"),
            rs.getString("condition"),
            rs.getString("status"),
            toIso(rs.getObject("listed_at")),
            rs.getInt("sold_count"),
            rs.getString("color")
        );
    }

    public static StoredProduct mapStoredProduct(ResultSet rs, String schema) throws SQLException {
        Product product = mapProduct(rs);
        return new StoredProduct(product, schema);
    }

    public static Purchase mapPurchase(ResultSet rs) throws SQLException {
        return new Purchase(
            rs.getString("id"),
            rs.getString("product_id"),
            rs.getString("product_name"),
            rs.getString("buyer_id"),
            rs.getString("seller_id"),
            toDouble(rs.getObject("amount")),
            toIso(rs.getObject("purchased_at")),
            rs.getString("status")
        );
    }

    public static Transaction mapTransaction(ResultSet rs) throws SQLException {
        return new Transaction(
            rs.getString("id"),
            rs.getString("type"),
            toDouble(rs.getObject("amount")),
            rs.getString("from_user_id"),
            rs.getString("to_user_id"),
            rs.getString("product_id"),
            rs.getString("description"),
            toIso(rs.getObject("created_at")),
            rs.getString("status")
        );
    }
}
